apps/aecore/src/aec_chain_state.erl

```diff
@@ -447,75 +448,97 @@ assert_micro_signature(PrevNode, Node) ->
+            case apply_and_store_state_trees(Node, TreesIn, ForkInfo, State) of
+                {ok, TreesOut, ForkInfoOut} ->
+                    update_next_state_tree(Node, TreesOut, ForkInfoOut, State);
                 error ->
-                    {State, Difficulty}
+                    {State, ForkInfo#fork_info.difficulty}
             end
     end.
 
-update_next_state_tree(Node, Trees, Difficulty, ForkId, State) ->
+update_next_state_tree(Node, Trees, ForkInfo, State) ->
     Hash = hash(Node),
     State1 = set_top_block_hash(Hash, State),
     case db_children(Node) of
-        [] -> {State1, Difficulty};
+        [] -> {State1, ForkInfo#fork_info.difficulty};
         [Child|Left] ->
             %% If there is only one child, it inherits the fork id.
             %% For more than one child, we neeed new fork_ids, which are
             %% the first node hash of each new fork.
-            Children = [{Child, ForkId}|[{C, hash(C)}|| C <- Left]],
-            update_next_state_tree_children(Children, Trees, Difficulty,
-                                            Difficulty, State1)
+            Children = [{Child, ForkInfo}|
+                        [{C, ForkInfo#fork_info{fork_id = hash(C)}}
+                         || C <- Left]
+                       ],
+            Difficulty = ForkInfo#fork_info.difficulty,
+            update_next_state_tree_children(Children, Trees, Difficulty, State1)
     end.
 
-update_next_state_tree_children([],_Trees,_Difficulty, Max, State) ->
+update_next_state_tree_children([],_Trees, Max, State) ->
     {State, Max};
-update_next_state_tree_children([{Child, ForkId}|Left], Trees, Difficulty, Max, State) ->
-    {State1, Max1} = update_state_tree(Child, Trees, Difficulty, ForkId, State),
+update_next_state_tree_children([{Child, ForkInfo}|Left], Trees, Max, State) ->
+    {State1, Max1} = update_state_tree(Child, Trees, ForkInfo, State),
     case Max1 > Max of
         true ->
-            update_next_state_tree_children(Left, Trees, Difficulty, Max1, State1);
+            update_next_state_tree_children(Left, Trees, Max1, State1);
         false ->
             State2 = set_top_block_hash(get_top_block_hash(State), State1),
-            update_next_state_tree_children(Left, Trees, Difficulty, Max, State2)
+            update_next_state_tree_children(Left, Trees, Max, State2)
     end.
 
 get_state_trees_in(Node, State) ->
     case node_is_genesis(Node, State) of
         true  ->
             {ok,
              aec_block_genesis:populated_trees(),
-             aec_block_genesis:genesis_difficulty(),
-             hash(Node)};
-        false -> db_find_state(prev_hash(Node))
+             #fork_info{ difficulty = aec_block_genesis:genesis_difficulty()
+                       , fork_id = hash(Node)
+                       , fees = 0
+                       }
+            };
+        false ->
+            PrevHash = prev_hash(Node),
+            case db_find_state(PrevHash) of
```

Comment:
Both db_find_state and db_get_node for same hash, synergy possible?

Reply:
It is from two different tables. Not much synergy to find there.